File: backend/data-transformer/src/models/feat.ts
import { Exclude, Expose, Transform } from "class-transformer";
import type { Publication } from "./secondary/publication";
import { mapToRule } from "./secondary/rule";
import { assignHeaderIds, cleanupHTML, stepHeaders } from "./utils/html";

export class Feat {
    @Expose({ name: "_id" })
    fvttId!: string;

    @Expose()
    name!: string;

    @Transform(({ obj }) => obj.system.actionType.value)
    @Expose()
    actionType!: string;

    @Transform(({ obj }) => obj.system.actions.value)
    @Expose()
    actions!: number;

    @Transform(({ obj }) => obj.system.category)
    @Expose()
    category!: string;

    @Transform(({ obj }) => {
        let description = obj.system.description.value;
        description = stepHeaders(description);
        description = assignHeaderIds(description);
        description = cleanupHTML(description);
        return description;
    })
    @Expose()
    description!: string;

    @Transform(({ obj }) => obj.system.level.value)
    @Expose()
    level!: number;

    @Transform(({ obj }) =>
        obj.system.prerequisites.value.map((v: any) => v.value),
    )
    @Expose()
    prerequisites!: string[];

    @Transform(({ obj }) => obj.system.publication)
    @Expose()
    publication!: Publication;

    // @Transform(({ obj }) => obj.system.rules.map((v: any) => mapToRule(v))) TODO: Check rules
    @Transform(({ obj }) => obj.system.rules)
    @Expose()
    rules!: any[];

    @Transform(({ obj }) => obj.system.traits.rarity)
    @Expose()
    rarity!: string;

    @Transform(({ obj }) => obj.system.traits.value)
    @Expose()
    traits!: string[];

    @Transform(({ obj }) => obj.system.slug)
    @Expose()
    slug!: string;

    @Exclude() system!: any;
    @Exclude() img!: any;
    @Exclude() _stats!: any;
    @Exclude() type!: any;
    @Exclude() effects!: any; // TODO: Check
}

File: backend/src/models/primary/feat.rs
use crate::{
    helpers::null_to_default,
    models::{LinkPreview, Summary, SummaryData, publication::Publication, rule::Rule},
    storeable::Storeable,
};
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Feat {
    // CompendiuancestrymFile
    #[serde(rename = "_id")]
    pub mongo_id: Option<ObjectId>,
    pub fvtt_id: String,
    pub name: String,
    pub action_type: String,
    pub actions: Option<u8>,
    pub category: String,
    pub description: String,
    pub level: u8,
    #[serde(default, deserialize_with = "null_to_default")]
    pub prerequisites: Vec<String>,
    pub publication: Publication,
    #[serde(default, deserialize_with = "null_to_default")]
    pub rules: Vec<Rule>,
    pub rarity: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub tags: Vec<String>,
    #[serde(default, deserialize_with = "null_to_default")]
    pub traits: Vec<String>,
    pub slug: String,
    #[serde(rename = "type")]
    pub data_type: String,
}

impl Storeable for Feat {
    fn table_name() -> &'static str {
        "feat"
    }
}

impl From<Feat> for Summary {
    fn from(value: Feat) -> Self {
        let data = vec![SummaryData::String {
            value: value.name.clone(),
            link: None,
            tooltip: None,
            abbreviation: None,
        }];

        Self {
            id: value.fvtt_id,
            name: value.name,
            slug: value.slug,
            data,
        }
    }
}

impl From<Feat> for LinkPreview {
    fn from(value: Feat) -> Self {
        Self {
            slug: value.slug,
            name: value.name,
            description: value.description,
            rarity: Some(value.rarity),
            traits: value.traits,
        }
    }
}
